packages/api-plugin-simple-schema/src/queries/introspectSchema.js

```diff
@@ -0,0 +1,75 @@
+import SimpleSchema, { toJsonSchema } from "simpl-schema";
+
+const currentPath = ["$"];
+const inputSchema = new SimpleSchema({
+  shopId: String,
+  schemaName: String
+});
+
+/**
+ * @name updateJsonSchema
+ * @summary Update the JSON Schema to add additional path field for each object
+ * @param {Object} jsonSchema - JSON Schema object
+ * @returns {Object} JSON Schema object with path field added
+ */
+function updateJsonSchema(jsonSchema) {
+  if (jsonSchema.type === "object" && jsonSchema.properties) {
+    const currentObject = jsonSchema.properties;
+    Object.keys(currentObject).forEach((key) => {
+      currentObject[key].path = `${currentPath.join(".")}.${key}`;
+      if (currentObject[key].type === "object") {
+        currentPath.push(key);
+        updateJsonSchema(currentObject[key]); // recursive call
+        currentPath.pop();
+      } else if (currentObject[key].type === "array" && currentObject[key].items && currentObject[key].items.length) {
+        currentPath.push(key);
+        currentPath.push("[0]");
+        const itemObject = currentObject[key].items[0];
+        updateJsonSchema(itemObject); // recursive call
+        itemObject.path = `${currentPath.join(".")}`;
+        currentPath.splice(-2, 2);
+      } else {
+        currentObject[key].path = `${currentPath.join(".")}.${key}`; // when not object/array, we just update the path
+      }
+    });
+  } else if (jsonSchema.type === "array" && jsonSchema.items && jsonSchema.items.length) {
+    currentPath.push("[0]");
+    const itemObject = jsonSchema.items[0];
+    updateJsonSchema(itemObject); // recursive call
+    itemObject.path = `${currentPath.join(".")}`;
+    currentPath.pop();
+  } else {
+    jsonSchema.path = currentPath.join("."); // when not object/array, we just update the path
+  }
+  jsonSchema.path = currentPath.join("."); // top level path
+  return jsonSchema;
+}
+
+/**
+ * @name introspectSchema
+ * @summary Extract the Schema of the required collection from context, convert to Json and return
+ * @param {Object} context - an object containing the per-request state
+ * @param {Object} args - an object of all arguments that were sent by the client
+ * @param {String} [args.shopId] - shopID
+ * @param {String} [args.schemaName] - schemaName to query
+ * @returns {Promise<Object>} JSON Schema object Promise
+ */
+export default async function introspectSchema(context, { shopId, schemaName } = {}) {
+  const validationString = "reaction:legacy:simpleSchema";
+  inputSchema.validate({ shopId, schemaName });
+
+  // Check for global schema permissions, if not found, check for schema specific permissions
+  try {
+    await context.validatePermissions(validationString, "introspect:*", { shopId });
+  } catch (error) {
+    await context.validatePermissions(validationString, `introspect:${schemaName}`, { shopId }); // if this throws, its valid access denied
```

Comment:
The first argument is the `resource` and when necessary, that can include a fourth piece that targets a specific resource by ID or name.

For example, we do this:

```js
await context.validatePermissions(`reaction:legacy:accounts:${accountId}`, "add:emails", {
  owner: account.userId
});
```

So it makes more sense to me to use `` `reaction:legacy:simpleSchema:${schemaName}` `` as the resource name and `"introspect"` as the action.

The only problem is that in `api-plugin-authorization-simple`, I don't think we ever properly coded `hasPermission` to work with that specific level of permissions. The other auth plugin supported it but not the simple auth.

The way it _should_ work is that you would not need the try/catch but would only do

```js
await context.validatePermissions(`reaction:legacy:simpleSchema:${schemaName}`, "introspect", { shopId });
```

And then if the schema were `cart` for example, the user would be allowed if they had _either_ `reaction:legacy:simpleSchema:cart` permission _or_ `reaction:legacy:simpleSchema` permission. If their permission does not have the specific `:cart` ending, then it means "all".

However, right now I think the simple auth plugin ignores the `:cart` piece and would give "all" access to everyone. I believe we only intended it to work that way as a transition but I guess it got stuck that way.

So... what you have can work as a workaround. Or you could take this opportunity to "fix" the simple auth permission checking.

Reply:
@aldeed cc: @zenweasel 
I checked the code in `hasPermissions` and as you mentioned it is ignoring the 4th part of the resource string which could be used for the resource specifier like ID. Meanwhile, there are few functions which are passing this 4th field to validatePermissions (like `addressBookAdd` and few others), but looks like this is getting discarded.  So if we introduce the validation for this 4th field, I am assuming that, these functions will suddenly start to fail with 'Access' issues.

So I think it would be better to deal with this entire auth plugin stuff separately in a new ticket instead of including this in introspectSchema PR. Please let me know if you disagree.

Reply:
@sujithvn I could be wrong, but I don't think they would start to fail because the changes should be additive. I mean that if a user has `1:2:3` permission, that should still be allowed for `validatePermissions("1:2:3:4")`. `1:2:3` can still mean `1:2:3:*` as it does now, but we would _additionally_ allow any user who has `1:2:3:4` listed.

If you want to do that in a later PR, which I agree would be good, then I think you'll just need to change `checkAndSeedDb` to add only a single permission, `reaction:legacy:simpleSchema/introspect`, which would mean "can introspect any schema".

Reply:
@aldeed 
what you said make sense. With the current update, we will extract both `1:2:3` and also `1:2:3:4` (if provided) versions of the permission strings  and  check it against the global & shop level permission (if shopId provided)

Also the permission `reaction:legacy:simpleSchema/introspect` is added in seed script & migration for shop-managet, owner & system-manager